src/backend/storage/smgr/smgr.c

```diff
@@ -175,6 +174,18 @@ smgrshutdown(int code, Datum arg)
 	}
 }
 
+#define SMGR_CHAIN_LOOKUP(SMGR_METHOD) while(chain_index < reln->smgr_chain.size && smgrsw[reln->smgr_chain.chain[chain_index]].SMGR_METHOD == NULL) chain_index++
+#define SMGR_CHAIN_CHECK_END_ASSERT() Assert(chain_index < reln->smgr_chain.size)
+
+void
+smgr_open_internal(SMgrRelation reln, SmgrChainIndex chain_index)
```

Comment:
It bothers me a bit that functions to be called from external entities have `_internal` in the name. Maybe `smgr_open_next`?

Plus it'd be nice to reference chain_index implicitly w/o passing it as an arg. But I can't come up with how) Maybe at least have some macros so user function don't have to write `chain_index + 1` all the time?

Reply:
"Plus it'd be nice to reference chain_index implicitly w/o passing it as an arg." - we can use macros, but that won't be pretty, I tried.

Or we can go with global state, but I wanted to avoid that, because if a storage manager ends up calling sql functions, like we did when we stored metadata in the catalog, we'll end up with multiple levels of smgr calls, and a global state would break - unless I also add handling for this, which would again complicate things.